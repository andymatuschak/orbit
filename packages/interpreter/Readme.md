# @withorbit/interpreter

This package, in conjunction with `@withorbit/ingester`, lets you import Orbit prompts from some other data source, and to keep your Orbit database in sync with that data source, ingesting missing prompts and deleting deleted ones as necessary.

This package is responsible for parsing those data sources and producing a JSON file which describes the prompts embedded within it. `ingester` can then synchronize an Orbit database with the prompts in that file. 

Right now, the only support data source is a folder of Markdown notes, so that you can commingle prompts with prose within a personal knowledge system.

## Example usage

⚠️ If you were a user of the older `@withorbit/note-sync` tool, see the migration notes at the bottom of this document before proceeding.

```
# Optional: set up a test note.
mkdir -p TestNotes
echo "# Test note\nThis is a {test cloze} note." > TestNotes/note.md

# Extract the prompts from a folder of Markdown notes and write a record of them to prompts.json.
# Here assuming you're at the orbit repo root!
bun run --cwd packages/interpreter interpret TestNotes prompts.json

# Sync a local Orbit database with that data source.
bun run --cwd packages/ingester ingest ~/myOrbitDB.orbitStore prompts.json 

# Optional: sync that local Orbit database with Orbit's server. (see documentation in ../sync/Readme.md for more)
ORBIT_TOKEN=myUserToken ORBIT_ENV=production bun run --cwd packages/sync sync ~/myOrbitDB.orbitStore
```

## Markdown prompt syntax

For example, here's an excerpt of some notes I wrote as I was studying Service Workers:

> During installation, service workers begin in the {installing} state, then transition to {activated} (or {error}) after {all the associated resources are fetched}. Before reaching that state, the service worker may transition to {waiting} until {no pre-existing service worker is controlling an open page}.
>
> Once activated, a service worker {performs one-time startup computation}, then transitions to {idle}. From that state, it’ll handle {fetch or message events} until it eventually terminates.
>
> Q. What’s the initial-registration gotcha for service workers’ control of web pages?
>
> A. They won’t control the web page which registered them until it’s refreshed, unless that client is specifically claimed.

### The basics

#### Creating typical two-sided SRS prompts

> Q. How many dimensions are in a qubit's vector space?
>
> A. Two.

The empty line between the question and answer is optional. The question and answer cannot currently span multiple paragraphs: the paragraph including `Q. ` or `A. ` is extracted as that field.

#### Creating cloze deletion prompts

In the context of prose notes, I'm finding cloze deletions are often somewhat more natural. This paragraph maps onto a single cloze deletion prompt with three cards:

> Once activated, a service worker {performs one-time startup computation}, then transitions to {idle}. From that state, it’ll handle {fetch or message events} until it eventually terminates.

The cloze prompt will use the entire paragraph surrounding the text. For example, this two-sided prompt is equivalent to one of the cards extracted from the previous example:

> Q. Once activated, a service worker ______, then transitions to idle. From that state, it’ll handle fetch or message events until it eventually terminates.
>
> A. Once activated, a service worker *performs one-time startup computation*, then transitions to idle. From that state, it’ll handle fetch or message events until it eventually terminates.

#### Idempotency and identity

This system is meant to operate idempotently. That is: you can keep revising your note files over time, and it'll keep track of changes accordingly. As you change your notes, the system will maintain your SRS state for all the embedded prompts, except for the prompts you've directly edited.

Somewhat more precisely, the embedded prompts have *identity*. You can modify the note around a prompt or even move the prompt to a new note, and your review history will be preserved. But if you modify a prompt's text, it will be treated as a new prompt, and your review history won't be ported from the old prompt. That's because this system's based on dumb plaintext files, which don't have enough semantic structure to unambiguously specify whether a given modification represents a new prompt or a modification of an old one. Fixing this would require introducing heuristics or extra identifying markup.

Two-sided prompt identities are derived from the hash of their question and answer text. Cloze prompt identities are derived from the hash of their containing paragraph.

While the system will happily track prompts if you rename notes or move prompts between note files, the behavior is undefined if identical prompts appear in multiple note files.

## Migrating from `note-sync`

`interpreter` and `ingester` replace the older `@withorbit/note-sync` package with a more maintainable and flexible implementation. **⚠️ Before using these tools**, you should make a copy of your Orbit database, then run this command to migrate it:

```
# (from this package's folder)
bun run migrateNoteSync ~/myOrbitDB.orbitStore
```

If you don't run this tool, the old `note-sync` prompts will stick around, but new duplicates will be added with no review history.

---

```
Copyright 2023 Andy Matuschak
SPDX-License-Identifier: Apache-2.0
```
