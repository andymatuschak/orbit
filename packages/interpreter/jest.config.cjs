module.exports = {
  moduleNameMapper: {
    "^(\\.{1,2}/.*)\\.js$": "$1",
  },
  testEnvironment: "node",
  testMatch: ["**/?(*.)+(spec|test).ts?(x)"],
  testPathIgnorePatterns: ["dist", "node_modules"],
  transformIgnorePatterns: [],
  transform: {
    "\\.[jt]s$": ["babel-jest", { cwd: __dirname }],
  },
};
