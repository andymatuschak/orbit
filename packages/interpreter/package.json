{
  "name": "@withorbit/interpreter",
  "version": "0.0.1",
  "type": "module",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "private": true,
  "sideEffects": false,
  "scripts": {
    "build": "tsc -b",
    "test": "jest",
    "interpret": "bun src/bin/run.ts",
    "migrateNoteSync": "node dist/bin/migrateNoteSync.js"
  },
  "dependencies": {
    "@withorbit/core": "0.0.1",
    "@withorbit/ingester": "0.0.1",
    "@withorbit/store-fs": "0.0.1",
    "mdast-util-from-markdown": "^2.0.0",
    "mdast-util-heading-range": "^4.0.0",
    "micromark-util-character": "^2.0.1",
    "micromark-util-symbol": "^2.0.0",
    "remark-frontmatter": "^5.0.0",
    "remark-math": "^6.0.0",
    "remark-parse": "^11.0.0",
    "remark-stringify": "^11.0.0",
    "remark-wiki-link": "^2.0.1",
    "unified": "^11.0.3",
    "unist-util-filter": "^5.0.1",
    "unist-util-map": "^4.0.0",
    "unist-util-parents": "^3.0.0",
    "unist-util-select": "^5.0.0"
  },
  "devDependencies": {
    "@babel/preset-env": "^7.16.11",
    "@types/jest": "^29.5.5",
    "@types/mdast": "^4.0.1",
    "@types/node": "^20.10.4",
    "@types/unist": "^3.0.0",
    "babel-jest": "^29.7.0",
    "jest": "^29.7.0",
    "micromark-util-types": "^2.0.0",
    "typescript": "^5.3.3"
  }
}
