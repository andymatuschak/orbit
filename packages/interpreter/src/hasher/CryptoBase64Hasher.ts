import {
  ClozeTaskContent,
  PlainTaskContent,
  QATaskContent,
  TaskContent,
  TaskContentType,
  TaskSpec,
} from "@withorbit/core";
import crypto from "crypto";
import { Hasher } from "./hasher.js";

// normalize each of the objects into an array of strings to ensure
// that the hash is not sensitive to key ordering
const taskContentDeterministicOrder: {
  [key in TaskContentType]: (spec: TaskContent) => (string | number | null)[];
} = {
  [TaskContentType.QA]: (content) => {
    const qa = content as QATaskContent;
    return [content.type, qa.body.text, qa.answer.text];
  },
  [TaskContentType.Cloze]: (content) => {
    const cloze = content as ClozeTaskContent;
    let output: (string | number | null)[] = [content.type, cloze.body.text];
    for (const [id, component] of Object.entries(cloze.components)) {
      output = output.concat(
        id,
        component.order,
        ...component.ranges.flatMap((r) => [r.hint, r.startIndex, r.length]),
      );
    }
    return output;
  },
  [TaskContentType.Plain]: (content) => {
    const plain = content as PlainTaskContent;
    return [content.type, plain.body.text];
  },
};

export const CryptoBase64Hasher: Hasher = {
  hash(spec: TaskSpec): string {
    let ordering = taskContentDeterministicOrder[spec.content.type](
      spec.content,
    );
    ordering = [spec.type, ...ordering];
    return crypto
      .createHash("sha256")
      .update(JSON.stringify(ordering))
      .digest("base64");
  },
};
