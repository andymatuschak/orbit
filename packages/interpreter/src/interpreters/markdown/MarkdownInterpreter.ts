import {
  ClozeTaskContentComponent,
  TaskContentType,
  TaskSpec,
  TaskSpecType,
} from "@withorbit/core";
import {
  Ingestible,
  IngestibleItem,
  IngestibleItemIdentifier,
  IngestibleSource,
  IngestibleSourceIdentifier,
} from "@withorbit/ingester";
import mdast, * as Mdast from "mdast";
import { selectAll } from "unist-util-select";
import { Hasher } from "../../hasher/hasher.js";
import { InterpretableFile, Interpreter } from "../../interpreter.js";
import {
  clozeNodeType,
  findAllPrompts,
  parseMarkdown,
  processor,
  Prompt,
} from "./markdown.js";
import { getNoteTitle } from "./utils/getNoteTitle.js";
import { getStableBearID } from "./utils/getStableBearID.js";

export class MarkdownInterpreter implements Interpreter {
  private _hasher: Hasher;
  constructor(hasher: Hasher) {
    this._hasher = hasher;
  }

  async interpret(files: InterpretableFile[]): Promise<Ingestible> {
    const nullableSources = await Promise.all(
      files.map(async (file): Promise<IngestibleSource | null> => {
        const root = await parseMarkdown(file.content);
        const bearId = getStableBearID(root);

        let identifier: IngestibleSourceIdentifier;
        let url: string | undefined;
        if (bearId) {
          identifier = bearId.id as IngestibleSourceIdentifier;
          url = bearId.openURL;
        } else {
          identifier = file.path as IngestibleSourceIdentifier;
        }

        const prompts = findAllPrompts(root);
        const noteTitle = getNoteTitle(root);

        return {
          identifier,
          title: noteTitle ?? file.name,
          url,
          items: prompts.map((prompt): IngestibleItem => {
            const spec = convertInterpreterPromptToIngestible(prompt);
            const identifier = this._hasher.hash(
              spec,
            ) as IngestibleItemIdentifier;
            return { identifier, spec };
          }),
        };
      }),
    );
    const sources = nullableSources.filter(
      (source) => source !== null,
    ) as IngestibleSource[];
    return { sources };
  }
}

function convertInterpreterPromptToIngestible(prompt: Prompt): TaskSpec {
  if (prompt.type === "qaPrompt") {
    return {
      type: TaskSpecType.Memory,
      content: {
        type: TaskContentType.QA,
        body: {
          text: processor
            .stringify(prompt.question as unknown as mdast.Root)
            .trimEnd(),
          attachments: [],
        },
        answer: {
          text: processor
            .stringify(prompt.answer as unknown as mdast.Root)
            .trimEnd(),
          attachments: [],
        },
      },
    };
  } else {
    // We've got to strip the braces out of the clozes.
    const markupWithBraces = processor
      .stringify(prompt.block as unknown as Mdast.Root)
      .trimEnd();
    const ast = processor.parse(markupWithBraces);
    const clozes = selectAll(clozeNodeType, ast);
    let markupWithoutBraces = "";
    const components: { [componentID: string]: ClozeTaskContentComponent } = {};
    clozes.forEach((cloze, i) => {
      if (i === 0) {
        markupWithoutBraces += markupWithBraces.slice(
          0,
          clozes[0].position!.start.offset!,
        );
      } else {
        markupWithoutBraces += markupWithBraces.slice(
          clozes[i - 1].position!.end.offset!,
          cloze.position!.start.offset!,
        );
      }
      markupWithoutBraces += markupWithBraces.slice(
        cloze.position!.start.offset! + 1,
        cloze.position!.end.offset! - 1,
      );
      components[i.toString()] = {
        order: i,
        ranges: [
          {
            hint: null,
            startIndex: cloze.position!.start.offset! - i * 2,
            length:
              cloze.position!.end.offset! - cloze.position!.start.offset! - 2,
          },
        ],
      };
    });
    markupWithoutBraces += markupWithBraces.slice(
      clozes.at(-1)!.position!.end.offset!,
    );

    return {
      type: TaskSpecType.Memory,
      content: {
        type: TaskContentType.Cloze,
        body: {
          text: markupWithoutBraces,
          attachments: [],
        },
        components,
      },
    };
  }
}
