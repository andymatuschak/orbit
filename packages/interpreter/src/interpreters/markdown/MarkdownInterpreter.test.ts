import { CryptoBase64Hasher } from "../../hasher/CryptoBase64Hasher.js";
import { InterpretableFile } from "../../interpreter.js";
import { MarkdownInterpreter } from "./MarkdownInterpreter.js";

it("interprets bear export file with mixed content", async () => {
  const file: InterpretableFile = {
    name: "Hello World.md",
    path: "/some/file/path/Hello World.md",
    content: `# Hello World

This is an example document which will contain both normal prompts and cloze prompts.
    
Q. What is the influence of A on B?
A. It makes B harder to read w.r.t A
    
This is another paragraph.
    
Q. This is another prompt.
A. Which contains a completely valid answer.
    
This is a {test} cloze prompt
    
<!-- {BearID:0DE4CB36-2EFC-4207-9074-667CBAE25ABB-12048-0000D97D373A46D2} -->`,
  };
  const interpreter = new MarkdownInterpreter(CryptoBase64Hasher);
  const ingestible = await interpreter.interpret([file]);
  expect(ingestible.sources).toHaveLength(1);
  expect(ingestible.sources[0]).toMatchSnapshot();
});

it("interprets raw markdown file with mixed content", async () => {
  const file: InterpretableFile = {
    name: "Hello World.md",
    path: "/some/file/path/Hello World.md",
    content: `# Hello World

This is an example document which will contain both normal prompts and cloze prompts.
    
Q. What is the influence of A on B?
A. It makes B harder to read w.r.t A
    
This is another paragraph.
    
Q. This is another prompt.
A. Which contains a completely valid answer.
    
This is a {test} cloze {prompt}`,
  };
  const interpreter = new MarkdownInterpreter(CryptoBase64Hasher);
  const ingestible = await interpreter.interpret([file]);
  expect(ingestible.sources).toHaveLength(1);
  expect(ingestible.sources[0]).toMatchSnapshot();
});
