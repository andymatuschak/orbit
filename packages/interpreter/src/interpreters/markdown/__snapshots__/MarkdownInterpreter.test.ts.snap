// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interprets bear export file with mixed content 1`] = `
{
  "identifier": "0DE4CB36-2EFC-4207-9074-667CBAE25ABB-12048-0000D97D373A46D2",
  "items": [
    {
      "identifier": "VGkBC83lHTUZs7mu7xxqJpvFuGys8iXzL54YDXgMB6E=",
      "spec": {
        "content": {
          "body": {
            "attachments": [],
            "text": "This is a test cloze prompt",
          },
          "components": {
            "0": {
              "order": 0,
              "ranges": [
                {
                  "hint": null,
                  "length": 4,
                  "startIndex": 10,
                },
              ],
            },
          },
          "type": "cloze",
        },
        "type": "memory",
      },
    },
    {
      "identifier": "HsSkb6EQxnC4lUPQ66v1yjarOgo5aSXQPZ3QNErckGw=",
      "spec": {
        "content": {
          "answer": {
            "attachments": [],
            "text": "It makes B harder to read w.r.t A",
          },
          "body": {
            "attachments": [],
            "text": "What is the influence of A on B?",
          },
          "type": "qa",
        },
        "type": "memory",
      },
    },
    {
      "identifier": "IolmfjQEMhQ7/rivRDy23KqggsnH9RI51WuR78zMcwg=",
      "spec": {
        "content": {
          "answer": {
            "attachments": [],
            "text": "Which contains a completely valid answer.",
          },
          "body": {
            "attachments": [],
            "text": "This is another prompt.",
          },
          "type": "qa",
        },
        "type": "memory",
      },
    },
  ],
  "title": "Hello World",
  "url": "bear://x-callback-url/open-note?id=0DE4CB36-2EFC-4207-9074-667CBAE25ABB-12048-0000D97D373A46D2",
}
`;

exports[`interprets raw markdown file with mixed content 1`] = `
{
  "identifier": "/some/file/path/Hello World.md",
  "items": [
    {
      "identifier": "jC4/8ovGX7TuYtKGkP6CZu42KteIbfW+UnFqP3WPong=",
      "spec": {
        "content": {
          "body": {
            "attachments": [],
            "text": "This is a test cloze prompt",
          },
          "components": {
            "0": {
              "order": 0,
              "ranges": [
                {
                  "hint": null,
                  "length": 4,
                  "startIndex": 10,
                },
              ],
            },
            "1": {
              "order": 1,
              "ranges": [
                {
                  "hint": null,
                  "length": 6,
                  "startIndex": 21,
                },
              ],
            },
          },
          "type": "cloze",
        },
        "type": "memory",
      },
    },
    {
      "identifier": "HsSkb6EQxnC4lUPQ66v1yjarOgo5aSXQPZ3QNErckGw=",
      "spec": {
        "content": {
          "answer": {
            "attachments": [],
            "text": "It makes B harder to read w.r.t A",
          },
          "body": {
            "attachments": [],
            "text": "What is the influence of A on B?",
          },
          "type": "qa",
        },
        "type": "memory",
      },
    },
    {
      "identifier": "IolmfjQEMhQ7/rivRDy23KqggsnH9RI51WuR78zMcwg=",
      "spec": {
        "content": {
          "answer": {
            "attachments": [],
            "text": "Which contains a completely valid answer.",
          },
          "body": {
            "attachments": [],
            "text": "This is another prompt.",
          },
          "type": "qa",
        },
        "type": "memory",
      },
    },
  ],
  "title": "Hello World",
  "url": undefined,
}
`;
